File: account-svc/src/main/java/club/hosppy/account/controller/AccountController.java
package club.hosppy.account.controller;

import club.hosppy.account.dto.*;
import club.hosppy.account.service.AccountService;
import com.github.structlog4j.ILogger;
import com.github.structlog4j.SLoggerFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import javax.validation.constraints.Min;
import java.util.List;

@RequiredArgsConstructor
@RestController
@RequestMapping("/accounts")
public class AccountController {

    private static final ILogger logger = SLoggerFactory.getLogger(AccountController.class);

    private final AccountService accountService;

    @GetMapping
    public List<AccountDto> listAccounts(@RequestParam(defaultValue = "0") @Min(0) int offset,
                                         @RequestParam(defaultValue = "10") @Min(0) int limit) {
        return accountService.list(offset, limit);
    }

    @GetMapping("/{userId}")
    public AccountDto getAccount(@PathVariable String userId) {
        return accountService.get(userId);
    }

    @PostMapping
    public AccountDto createAccount(@RequestBody @Valid CreateAccountRequest request) {
        return accountService.create(request.getName(), request.getEmail(), request.getPhoneNumber());
    }

    @PutMapping("/password")
    public void updatePassword(@RequestBody @Valid UpdatePasswordRequest request) {
        accountService.updatePassword(request.getUserId(), request.getPassword());
    }

    @PostMapping("/email")
    public void changeEmail(EmailConfirmation request) {
        accountService.changeEmailAndActivateAccount(request.getEmail(), request.getUserId());
    }

    @GetMapping("/activate/{token}")
    public void verifyActivateToken(@PathVariable String token) {
        accountService.verifyActivateToken(token);
    }

    @PostMapping("/activate/{token}")
    public void activateAccount(@PathVariable String token, @RequestBody @Valid ActivateAccountRequest request) {
        accountService.activateAccount(token, request.getPassword());
    }
}

File: src/main/kotlin/club/hosppy/account/controller/AccountController.kt
package club.hosppy.account.controller

import club.hosppy.account.dto.*
import club.hosppy.account.service.AccountService
import com.github.structlog4j.SLoggerFactory
import org.springframework.web.bind.annotation.*
import javax.validation.Valid
import javax.validation.constraints.Min

@RestController
@RequestMapping("/accounts")
class AccountController(
    private val accountService: AccountService,
) {

    @GetMapping
    fun listAccounts(
        @RequestParam(defaultValue = "0") page: @Min(0) Int,
        @RequestParam(defaultValue = "10") size: @Min(0) Int,
    ): List<AccountDto?>? {
        return accountService.list(page, size)
    }

    @GetMapping("/{email}")
    fun getAccount(@PathVariable email: String?): AccountDto? {
        return accountService.getByEmail(email)
    }

    @PostMapping
    fun createAccount(@RequestBody @Valid request: CreateAccountRequest): AccountDto? {
        return accountService.create(request.name, request.email, request.phoneNumber, request.password)
    }

    @PutMapping
    fun updateAccount(@RequestBody @Valid request: UpdateAccountRequest): AccountDto {
        // TODO get current userId
        val newAccountDto = AccountDto().apply {
            name = request.name
            avatarUrl = request.avatarUrl
            phoneNumber = request.phoneNumber
        }
        return accountService.update(newAccountDto)
    }

    @PutMapping("/password")
    fun updatePassword(@RequestBody request: @Valid UpdatePasswordRequest) {
        accountService.updatePassword(request.userId, request.password)
    }

    @PostMapping("/email")
    fun changeEmail(request: EmailConfirmation) {
        accountService.changeEmailAndActivateAccount(request.email, request.userId)
    }

    @GetMapping("/activate/{token}")
    fun activateAccount(@PathVariable token: String) {
        accountService.activateAccount(token)
    }

    companion object {
        private val logger = SLoggerFactory.getLogger(AccountController::class.java)
    }
}